### ISSUE TITLE
Discussion around ownership of pins / moving pins between tables

### ISSUE THREAD

**Issue by RaiLTC** [nbusseneau/BetterCartographyTable#10]
As the title says specifically, every single pin including the bosses pins, is gonna be deleted after the table is destroyed
I have not thoroughly tested it, but it did happen unfortunately

I have made every pin public before destroying the table except the ones I couldn't (EpicLoot bounties, bosses etc)

**Comment by nbusseneau** [comment]
All pins shared to a table are stored in the table's ZDO so this is expected. Can you expand on the use case, what were you trying to achieve? What were your expectations about how the cartography table should work, and in particular why you expected the table pins to carry over to future tables?

**Comment by RaiLTC** [comment]
I didn't like the location of the table and tried to move it
Expected behavior was having all of my pins become private and not wiped :p (especially boss pins, not sure why those were deleted)

You should at least have a warning regarding this because it seems really inconvenient. Having to private everything before moving a table seems like a ton of work later into the game and stuff considering I didn't see a private / public all button

**Comment by nbusseneau** [comment]
OK, it looks to me that your expectation here is that when you share something to the table, you still retain ownership of the pins you had shared. For example if you had shared boss pins to the table they became owned by the table, and not you, but I can see why that might be confusing if you expect the pins to still be yours even after sharing them to the table (which is not how the mod works at the moment).

There is already a warning for when you try to switch between a table between public and guild mode, because of the same thing happening if you do it (i.e. table pins disappear), I agree that adding a similar warning for when trying to destruct a table with pins would be good, I'll see what I can do 😄

But this does not address the core point of you expecting pins still to be "yours". This also ties back to your point about not having a "private all / public all" mechanism: can you expand on why you think that is useful? The point of the mod was to offer a way for communal cartography tables to be viable, i.e. have a place where multiple players can come and share pins (typically this cartography table would never need to be moved, because it would not be in any individual's base). However its goal is not for players to share all their pins at all times: there already are other mods that allow for that, and my goal here was to differentiate by covering the use case of "we share some stuff and sometimes play together, but we also want to do our own thing and not share everything at all times".

So rephrasing, my question is what is your use case in your group of friends, and why did you specifically choose this mod for it instead of another mod (e.g. one of those that allows sharing all pins at all times)? Depending on if I think that's sensible for Better Cartography Table I might think about adding more functionality or changing more stuff (e.g. trying to address the expectation that pins are still "yours").

**Comment by nbusseneau** [comment]
Added a warning popup when trying to destroy tables with any pins via 4fdf8c193e87d5a78b1a69fa04b16489c875d12c, released in [0.5.7](https://github.com/nbusseneau/BetterCartographyTable/releases/tag/0.5.7).

At least this will avoid this scenario taking someone else by surprise 😅

**Comment by RaiLTC** [comment]
Ah I see the point of the mod now!
I don't plan on changing your goal with the mod, so just having the warning about the table destruction would suffice!

And regarding the private all / public all thing, the point was to simply make it save time, because if you have a lot of pins and you wanted to share a lot, sometimes making them all public and then privating a few would be faster than making pins public one by one :p

**Comment by nbusseneau** [comment]
> And regarding the private all / public all thing, the point was to simply make it save time, because if you have a lot of pins and you wanted to share a lot, sometimes making them all public and then privating a few would be faster than making pins public one by one :p

Yeah I dig that. The thing is I didn't really expect anyone would have a need for that because there already are other mods that allow for sharing all pins at all times, contrary to how this mod stores pins on tables (similar to how vanilla does). This is because the goal is not for everything to be available at all times, and instead force players to come and interact with a communal table on a regular basis (and potentially even with 2 distinct tables: one public and one for their guild).

That's why I'm asking questions on your use case, I'm trying to figure out why you elected to use the mod instead of another one, and if I can make your life easier without changing the mod's goal. For example I'm currently thinking on ideas to "transport" a table somewhere else without destroying it (and thus without losing the pins). So yeah, if you have some time to share about how you and your group use the tables / pin sharing, that would help me think about what can benefit your use case 😄

**Comment by RaiLTC** [comment]
We play on a private server with a few people (around 6~), so we don't really use the guild feature

And the reason we use your mod is because it was the first one we saw with a respectable amount of downloads (meaning we can trust it being good :D), without looking into too much details as the title of the mod simply states that it makes the table better

As for making the mod better, simply adding the destruction warning and the planned transport feature will make it perfect. We'll have to look into downloading a mod that makes all the pins public or private. Would prefer if you had this functionality within the mod of course but can't ask for everything considering the idea you have behind this mod

Thanks for everything so far :p

EDIT:
Now that I think about it..

How about having an option within the table to export the pins onto an item say.. something like paper, which you are able to use on a different table to import the pins? Would be neat

**Comment by RaiLTC** [comment]
Now that I think about it..

How about having an option within the table to export the pins onto an item something like paper, which you are able to use on a different table to import the pins? Would be neat

**Comment by nbusseneau** [comment]
Yeah, using an item to transport the pins was what I was thinking as well 😄 

As for your use case, the mod I was thinking about that shares all pins at all times is this one: https://thunderstore.io/c/valheim/p/Mydayyy/ServerSideMap/

**Comment by nbusseneau** [comment]
Playing Valheim again lately so I'm reviving this because I've got an idea for how to transfer pins between tables:
- Each table has its own unique ID when created.
- On first sync with the table, we store the table ID alongside the pins.
- As long as you interact with the same table ID, it just works like now.
- If try to sync with a table with a different ID, you get a warning and you must choose one option:
  1. Replace your pins with pins from the table (effectively what is happening right now). This is probably what you would choose in situations where you switch fellowship, e.g. you played with some Guild before and switched to a new Guild.
  2. Replace the pins from the table with your own pins. This is probably what you would choose in situations where you migrate somewhere else and want to move the table with you (i.e. the original use case of this issue).

For option 2., I wonder whether it makes sense to only allow this if the table is empty, and if we should try to merge rather than replace. The latter would technically permit "pin stealing", i.e. if a spy switches guilds they can leak the former guild's pins, which might be a fun thing for RP servers.

**Comment by nbusseneau** [comment]
Initial implementation at c0355674fe80501a8a459449d107ff7d569b5305.

**Comment by nbusseneau** [comment]
Released in [0.8.0](https://github.com/nbusseneau/BetterCartographyTable/releases/tag/0.8.0).
